File: src/utils/writeRes.ts
import fs from "fs";
import { SolveResult } from "./DpllCenter";

const writeRes: (path: string, res: SolveResult) => void = (path, res) => {
  const wr = fs.createWriteStream(path);
  switch (res.state) {
    case -1:
      wr.write("s 0\n");
      break;
    case 0:
      wr.write("s -1\n");
      break;
    case 1:
      wr.write("s 1\n");
      break;
    default:
      break;
  }
  if (res.state === 1) {
    let s = "v";
    for (let i = 1; i < res.lits.length; i++) {
      s += " " + (res.lits[i] > 0 ? i : -i);
    }
    s += "\n";
    wr.write(s);
  }
  wr.write(`t ${res.time}`);
  wr.close();
};

export default writeRes;

File: addon/src/writeRes.cc
//
// Created by panda on 2020/2/22.
//
#include "writeRes.hh"
namespace final {
auto writeRes(State s, double time, Literals &lits, std::string path) -> void {
    std::ofstream out(path, std::ios::out | std::ios::trunc);
    out.setf(std::ios::fixed);
    out.unsetf(std::ios::floatfield);
    switch (s) {
    case satisfied:
        out << "s 1" << std::endl;
        break;
    case unsatisfied:
        out << "s 0" << std::endl;
        break;
    case unknown:
        out << "s -1" << std::endl;
        break;
    default:
        break;
    }
    if (s == satisfied) {
        out << "v";
        for (int i = 1; i < lits.size(); i++) {
            out << " " << ((lits[i].state == satisfied) ? i : -i);
        }
        out << std::endl;
    }
    out << "t " << time << std::endl;
    out.close();
}
} // namespace final